### PR TITLE
Vectorize conversions to BFloat16 on CPU

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80906

This adds explicit vectorization for converting float or double to
bfloat16. Most conversions are sufficiently handled by the
auto-vectorizer, but these conversions aren't (presumably due to
branching in the scalar conversion code).

Benchmark results with 512K elements on an AVX2 machine:

| conversion          | Before (us) | After (us) |
|---------------------|-------------|------------|
| float32 -> bfloat16 | 53.3        | 39.8       |
| float64 -> bfloat16 | 92.1        | 78.2       |

### CODE REVIEW

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h [hunk 1]
```diff
@@ -26,7 +26,7 @@ static inline void cvtbf16_fp32(const __m128i& a, __m256& o) {
 }
 
 static inline void cvtbf16_fp32(const __m256i& a, __m256& o1, __m256& o2) {
-  __m128i lo = _mm256_extractf128_si256(a, 0);
+  __m128i lo = _mm256_castsi256_si128(a);
```
Comment (ngimel) [inline aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:29]: why this change?

**Review by ngimel — COMMENTED** [review]
Do we have specific tests for these conversions, or are all the combinations of datatypes tested somewhere?

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:29]: This cast is an equivalent operation with zero-latency so it makes sense to use it instead. However the compiler almost certainly makes this change already so I can drop it if you think it's less clear.

**Comment by peterbell10** [comment]
I can find OpInfo tests for all of the `x.bfloat16()` style cast functions, however I'm not sure to what degree these tests correctness.

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:29]: Yeah for people who don't have these instructions memorized the previous approach is easier to understand

**Comment by peterbell10** [comment]
@ngimel PTAL

**Comment by peterbell10** [comment]
@ngimel ping

**Review by ngimel — APPROVED** [review]
